iot-hub/Tutorials/FirmwareUpdate/SimulatedDevice.js

```diff
@@ -0,0 +1,308 @@
+
+          sendStartingUpdate(fwUpdateDesiredProperties.fwVersion);
+          initiateFirmwareUpdateFlow(function(err, result) {
+            fwUpdateInProgress = false;
+            if (!err) {
+              console.log(chalk.green('Completed firmwareUpdate flow. New version: ' + result));
+              sendFinishedUpdate(result);
+            }
+           }, twin.properties.reported.firmware.currentFwVersion);
+        });
+        // </initiateUpdate>
+      }
+    });
+    console.log(chalk.green('Client connected to IoT Hub. Waiting for desired properties update.'));
+  }
+});
+
+// <firmwareupdateflow>
+// Implementation of firmwareUpdate flow
+function initiateFirmwareUpdateFlow(callback, currentVersion) {
+
+  async.waterfall([
+    downloadImage,
+    verifyImage,
+    applyImage,
+    reboot
+  ], function(err, result) {
+    if (err) {
+      console.error(chalk.red('Error occured firmwareUpdate flow : ' + err.message));
+      sendStatusUpdate('error', err.message);
+      setTimeout(function() {
+        console.log('Simulate rolling back update due to error');
+        sendStatusUpdate('rolledback', 'Rolled back to: ' + currentVersion);
+        callback(err, result);
+      }, 5000);
+    } else {
+      callback(null, result);
+    }
+  });
+}
+// </firmwareupdateflow>
+
+// <downloadimagephase>
+// Simulate downloading an image
+function downloadImage(callback) {
+  console.log('Simulating image download from: ' + desiredFirmwareProperties.fwPackageURI);
+
+  async.waterfall([
+    function(callback) {
+      sendStatusUpdate('downloading', 'Start downloading');
+      callback(null);
+    },
+    function(callback) {
+      // Simulate a delay downloading the image.
+      setTimeout(function() {
+        // Simulate some firmware image data
+        var imageData = '[Fake firmware image data]';
+        callback(null, imageData); 
+      }, 30000);
+    },
+    function(imageData, callback) {
+      console.log('Downloaded image data: ' + imageData);
+      sendStatusUpdate('downloading', 'Finished downloading');
+      callback(null, imageData);
+    }
+  ], function (err, result) {
+    callback(err, result);
+  });
+}
+// </downloadimagephase>
+
+// <verifyimagephase>
+// Simulate verifying an image
+function verifyImage(imageData, callback) {
+  console.log('Simulating image verification using checksum: ' + desiredFirmwareProperties.fwPackageCheckValue);
+
+  async.waterfall([
+    function(callback) {
+      sendStatusUpdate('verifying', 'Start verifying download');
+      callback(null);
+    },
+    function(callback) {
+      // Simulate verifying the image.
+      setTimeout(function() {
+        // Simulate verifying image data
+        callback(null);
+        // Simulate an error verifying the image data
+        //callback(new Error('Image verification failed'));
+      }, 10000);
+    },
+    function(callback) {
+      sendStatusUpdate('verifying', 'Finished verifying download');
```

Comment:
>verifying [](start = 24, length = 9)

that is where we'd use the checksum - it'd be better to have a real example of how to do that.